grpclb/src/main/java/io/grpc/grpclb/RoundRobinServerList.java

```diff
@@ -0,0 +1,118 @@
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.grpclb;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.TransportManager;
+import io.grpc.internal.ClientTransport;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Manages a list of server addresses to round-robin on.
+ */
+// TODO(zhangkun83): possibly move it to io.grpc.internal, as it can also be used by the round-robin
+// LoadBalancer.
+@ThreadSafe
+class RoundRobinServerList {
+  private final TransportManager tm;
+  private final List<EquivalentAddressGroup> list;
+  private final Iterator<EquivalentAddressGroup> cyclingIter;
+
+  private RoundRobinServerList(TransportManager tm, List<EquivalentAddressGroup> list) {
+    this.tm = tm;
+    this.list = list;
+    this.cyclingIter = Iterables.cycle(list).iterator();
+  }
+
+  ListenableFuture<ClientTransport> getTransportForNextServer() {
+    EquivalentAddressGroup currentServer;
+    synchronized (cyclingIter) {
+      // TODO(zhangkun83): receive transportShutdown and transportReady events, then skip addresses
+      // that have been failing.
+      currentServer = cyclingIter.next();
+    }
+    if (currentServer == null) {
+      // TODO(zhangkun83): drop the request by returnning a fake transport that would fail
+      // streams.
+      throw new UnsupportedOperationException("server dropping not implemented yet");
+    }
+    return Preconditions.checkNotNull(tm.getTransport(currentServer),
+        "TransportManager returned null for %s", currentServer);
+  }
+
+  @VisibleForTesting
+  List<EquivalentAddressGroup> getList() {
```

Comment:
It's this method that concerns me.  Even though its marked for testing, anyone who called this method would get a mutable copy of the list.